### PR TITLE
thrift proxy: support outlier

### PR DESCRIPTION
**Pull request by LawlietLi**
Signed-off-by: Yiheng Li <lyhutopi@gmail.com>
Commit Message: thrift proxy support outlier
Additional Description: Put upstream result to outlier monitor in thrift route filter.
Risk Level: Low, no behavior change
Testing: UTs
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc [hunk 1]
```diff
@@ -362,13 +362,19 @@ void Router::onUpstreamData(Buffer::Instance& data, bool end_stream) {
...
         } else {
+          upstream_request_->upstream_host_->outlierDetector().putResult(
+              Upstream::Outlier::Result::ExtOriginRequestFailed);
           incClusterScopeCounter({upstream_resp_reply_error_});
         }
         break;
 
       case MessageType::Exception:
+        upstream_request_->upstream_host_->outlierDetector().putResult(
+            Upstream::Outlier::Result::ExtOriginRequestFailed);
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc:377]: Should we handle the reset and end stream cases below? They imply something went wrong decoding the response.

**Review by zuercher — COMMENTED** [review]
Thanks! This looks generally correct to me.

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc (hunk 1 above)
Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:377]: +1

**Review by rgs1 — COMMENTED** [review]
LGTM once Stephan's commented is addressed. Thanks!

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc (hunk 1 above)
Comment (LawlietLi) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:377]: I've added reporting to `onResetStream`, which should handle the `end_stream` case. I'm not sure what result should be reported in the case of reset, does `ExtOriginRequestFailed` make sense?

Comment (zuercher) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:377]: I think so. It's an early termination of a response.

Comment (LawlietLi) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc:377]: ok，I've added code to handle this

**Review by zuercher — APPROVED** [review]

**Comment by zuercher** [comment]
The code looks good. Needs release notes (add an entry in `docs/root/version_history/current.rst`). I think perhaps in the minor behavior changes -- if someone has outlier detection enabled on a thrift cluster this change may change behavior for them.

Also, a note in `docs/root/intro/arch_overview/upstream/outlier.rst ` about outlier detection being available in thrift.

**Comment by LawlietLi** [comment]
I've added the doc, but I'm not a native speaker, so please help me out if there are any problems.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -105,6 +105,7 @@ New Features
 * router: added flag ``suppress_grpc_request_failure_code_stats`` to :ref:`key <envoy_v3_api_msg_extensions.filters.http.router.v3.Router>` to allow users to exclude incrementing HTTP status code stats on gRPC requests.
 * tcp: added support for :ref:`preconnecting <v1.18.0:envoy_v3_api_msg_config.cluster.v3.Cluster.PreconnectPolicy>`. Preconnecting is off by default, but recommended for clusters serving latency-sensitive traffic.
 * thrift_proxy: added per upstream metrics within the :ref:`thrift router <envoy_v3_api_msg_extensions.filters.network.thrift_proxy.router.v3.Router>` for request and response size histograms.
+* thrift_proxy: added :ref:`outlier <arch_overview_outlier_detection>` to support.
```
Comment (zuercher) [inline docs/root/version_history/current.rst:108]:
```suggestion
* thrift_proxy: added support for :ref:`outlier detection <arch_overview_outlier_detection>`.
```

**Review by zuercher — COMMENTED** [review]
Thanks!

**Review by zuercher — APPROVED** [review]

**Pull request merged**